polygeist/tools/cgeist/Lib/ValueCategory.cc

```diff
@@ -142,21 +142,19 @@ void ValueCategory::store(mlir::OpBuilder &builder, mlir::Value toStore) const {
 
     if (toStore.getType() != pt.getElementType()) {
       if (auto mt = toStore.getType().dyn_cast<MemRefType>()) {
-        if (auto spt =
-                pt.getElementType().dyn_cast<mlir::LLVM::LLVMPointerType>()) {
-          CGEIST_WARNING({
-            if (mt.getElementType() != spt.getElementType()) {
-              // llvm::errs() << " func: " <<
-              // val.getDefiningOp()->getParentOfType<FuncOp>() << "\n";
-              llvm::WithColor::warning()
-                  << "potential store type mismatch:\n"
-                  << "val: " << val << " tosval: " << toStore << "\n"
-                  << "mt: " << mt << "spt: " << spt << "\n";
-            }
-          });
-          toStore =
-              builder.create<polygeist::Memref2PointerOp>(loc, spt, toStore);
-        }
+        auto spt = pt.getElementType().cast<mlir::LLVM::LLVMPointerType>();
+        CGEIST_WARNING({
+          if (mt.getElementType() != spt.getElementType()) {
+            // llvm::errs() << " func: " <<
+            // val.getDefiningOp()->getParentOfType<FuncOp>() << "\n";
+            llvm::WithColor::warning()
+                << "potential store type mismatch:\n"
+                << "val: " << val << " tosval: " << toStore << "\n"
+                << "mt: " << mt << "spt: " << spt << "\n";
+          }
+        });
+        toStore =
+            builder.create<polygeist::Memref2PointerOp>(loc, spt, toStore);
```

Comment:
If the `dyn_cast` failed, we were doing nothing, generating wrong code and not reporting. Now we do report.

Reply:
This lead to the bad code in the lambda capture reported by @whitneywhtsang

Reply:
This change causes test suite regression, 250 passed -> 244 passed.

Reply:
Code was "correct" back then (we had an ABI violation due to bad typing of Boolean scalar values). This PR caused a silent error. I'm simply exposing the error here. If this is causing regressions, I suggest we first fix lambda capture in a different PR and then we can proceed with the fix in this PR.